src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.cs

```diff
@@ -0,0 +1,60 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.Security.Cryptography.Encoding.Tests.Cbor
+{
+    internal ref partial struct CborValueReader
+    {
+        private ReadOnlySpan<byte> _buffer;
+
+        public CborValueReader(ReadOnlySpan<byte> buffer)
+        {
+            _buffer = buffer;
+        }
+
+        private CborDataItem ReadInitialByte()
+        {
+            if (_buffer.IsEmpty)
+            {
+                throw new FormatException("end of buffer");
+            }
+
+            return new CborDataItem(_buffer[0]);
+        }
+
+        private CborDataItem ReadInitialByte(CborMajorType expectedType)
+        {
+            if (_buffer.IsEmpty)
+            {
+                throw new FormatException("end of buffer");
+            }
+
+            CborDataItem result = new CborDataItem(_buffer[0]);
+
+            if (expectedType != result.MajorType)
+            {
+                throw new ArgumentException("Major type does not match expected type");
```

Comment:
It looks like this is used as state validation for things like `ReadByteString(CborMajorType.ByteString)`; so InvalidOperationException.  (And tests :smile:)